src/pybind/mgr/dashboard_v2/services/ceph_service.py

```diff
@@ -53,8 +133,96 @@ def get_service(cls, service_name, service_id):
+        pool_stats = {}
+        rbdctx = rbd.RBD()
+        for pool_name in pool_names:
+            logger.debug("Constructing IOCtx %s", pool_name)
+            try:
+                ioctx = cls.mgr.rados.open_ioctx(pool_name)
+            except TypeError:
+                logger.exception("Failed to open pool %s", pool_name)
+                continue
+
+            mirror_mode = rbd.RBD_MIRROR_MODE_DISABLED
+            try:
+                mirror_mode = rbdctx.mirror_mode_get(ioctx)
+            except:  # noqa pylint: disable=W0702
+                logger.exception("Failed to query mirror mode %s", pool_name)
+
+            stats = {}
+            if mirror_mode == rbd.RBD_MIRROR_MODE_DISABLED:
+                continue
+            elif mirror_mode == rbd.RBD_MIRROR_MODE_IMAGE:
+                mirror_mode = "image"
+            elif mirror_mode == rbd.RBD_MIRROR_MODE_POOL:
+                mirror_mode = "pool"
+            else:
+                mirror_mode = "unknown"
+                stats['health_color'] = "warning"
+                stats['health'] = "Warning"
+
+            pool_stats[pool_name] = dict(stats, **{
+                'mirror_mode': mirror_mode
+            })
+
+        for daemon in daemons:
+            for _, pool_data in daemon['status'].items():
+                stats = pool_stats.get(pool_data['name'], None)
+                if stats is None:
+                    continue
+
+                if pool_data.get('leader', False):
+                    # leader instance stores image counts
+                    stats['leader_id'] = daemon['metadata']['instance_id']
+                    stats['image_local_count'] = pool_data.get('image_local_count', 0)
+                    stats['image_remote_count'] = pool_data.get('image_remote_count', 0)
+
+                if (stats.get('health_color', '') != 'error' and
+                        pool_data.get('image_error_count', 0) > 0):
+                    stats['health_color'] = 'error'
+                    stats['health'] = 'Error'
+                elif (stats.get('health_color', '') != 'error' and
+                      pool_data.get('image_warning_count', 0) > 0):
+                    stats['health_color'] = 'warning'
+                    stats['health'] = 'Warning'
+                elif stats.get('health', None) is None:
+                    stats['health_color'] = 'success'
+                    stats['health'] = 'OK'
+
+        for _, stats in pool_stats.items():
+            if stats.get('health', None) is None:
+                # daemon doesn't know about pool
+                stats['health_color'] = 'error'
+                stats['health'] = 'Error'
+            elif stats.get('leader_id', None) is None:
+                # no daemons are managing the pool as leader instance
+                stats['health_color'] = 'warning'
+                stats['health'] = 'Warning'
+        return pool_stats
+
+    @classmethod
+    @ViewCache()
```

Comment:
the functions in these mixin classes shouldn't use the `@ViewCache` decorator. The `@ViewCache` decorator should be use solely in controller classes so that it's explicit in the controller code which data is being cached, otherwise when I want to re-use this code I may not now that it's already being cached, and worst, I might not want it to be cached.

Reply:
Does that completely exclude Mixins to be decorated by the `@ViewCache` decorator?

Reply:
only in the mixin function declaration (mixin file). These mixin functions can, and some should, be used in controllers' functions decorated with `@ViewCache`.